x-pack/legacy/plugins/lens/public/indexpattern_plugin/filter_ratio.ts

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { i18n } from '@kbn/i18n';
+import { ExpressionFunction } from 'src/legacy/core_plugins/interpreter/types';
+import { KibanaDatatable } from '../types';
+
+interface FilterRatioKey {
+  id: string;
+}
+
+export const calculateFilterRatio: ExpressionFunction<
+  'lens_calculate_filter_ratio',
+  KibanaDatatable,
+  FilterRatioKey,
+  KibanaDatatable
+> = {
+  name: 'lens_calculate_filter_ratio',
+  type: 'kibana_datatable',
+  help: i18n.translate('xpack.lens.functions.calculateFilterRatio.help', {
+    defaultMessage: 'A helper to collapse two filter ratio rows into a single row',
+  }),
+  args: {
+    id: {
+      types: ['string'],
+      help: i18n.translate('xpack.lens.functions.calculateFilterRatio.id.help', {
+        defaultMessage: 'The column ID which has the filter ratio',
+      }),
+    },
+  },
+  context: {
+    types: ['kibana_datatable'],
+  },
+  fn(data: KibanaDatatable, { id }: FilterRatioKey) {
+    const newRows: KibanaDatatable['rows'] = [];
+
+    if (data.rows.length === 0) {
+      return data;
+    }
+
+    const entries = Object.entries(data.rows[0]);
+    const [[valueKey]] = entries.filter(([key]) => key.includes('filter-ratio'));
+
+    for (let i = 0; i < data.rows.length; i += 2) {
+      const row1 = data.rows[i];
+      const row2 = data.rows[i + 1];
+
+      const calculatedRatio = row2[valueKey]
```

Comment:
Per the "undefined" comment above, I'd be more comfortable if this was a defensive check: `row2 && row2[valueKey] ? ... : ...`

Reply:
It still doesn't look like we're handling the row2 undefined case.

Reply:
I added a check on line 45 to throw if the number of rows is odd. I _think_ that's enough, but maybe there's something I'm not defending against enough?